migrations/2025-03-06-060640_semver_ord/up.sql

```diff
@@ -0,0 +1,109 @@
+-- Add `semver_ord(num)` function to convert a semver string into a JSONB array for version comparison purposes.
+
+create or replace function semver_ord(num varchar) returns jsonb as $$
+declare
+    -- We need to ensure that the prerelease array has the same length for all
+    -- versions since shorter arrays have lower precedence in JSONB. We store
+    -- the first 10 parts of the prerelease string as pairs of booleans and
+    -- numbers or text values, and then a final text item for the remaining
+    -- parts.
+    max_prerelease_parts constant int := 10;
+
+    -- We ignore the "build metadata" part of the semver string, since it has
+    -- no impact on the version ordering.
+    match_result text[] := regexp_match(num, '^(\d+).(\d+).(\d+)(?:-([0-9A-Za-z\-.]+))?');
+
+    prerelease jsonb;
+    prerelease_parts text[];
+    prerelease_part text;
+    i int := 0;
+begin
+    if match_result is null then
+        return null;
+    end if;
+
+    if match_result[4] is null then
+        -- A JSONB object has higher precedence than an array, and versions with
+        -- prerelease specifiers should have lower precedence than those without.
```

Comment:
I went to check that we're not relying on any undocumented behaviour, and found [this gem](https://www.postgresql.org/docs/17/datatype-json.html#JSON-INDEXING):

> The btree ordering for jsonb datums is seldom of great interest[...]

Little did they know!

Reply:
yep, see https://github.com/rust-lang/crates.io/pull/10763#discussion_r1983818620 :)